src/main/java/hudson/remoting/AsyncFutureImpl.java

```diff
@@ -78,13 +79,27 @@ public synchronized V get() throws InterruptedException, ExecutionException {
         return value;
     }
 
+    @CheckForNull
     public synchronized V get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
-        if(!completed)
-            wait(unit.toMillis(timeout));
+        // The accuracy of wait(long) operation is milliseconds anyway, but ok.
+        long endWaitTime = System.nanoTime() + unit.toNanos(timeout);
+        while (!completed) {
+            long timeToWait = endWaitTime - System.nanoTime();
+            if (timeout < 0) {
```

Comment:
`s/timeout/timeToWait/`

Reply:
oh yes :( It makes me think that some test coverage could be actually useful